### PR TITLE
Accelerate LLaMA model loading

### PR DESCRIPTION
**Pull request by JF-D**
This PR is for accelerating LLaMA model weights loading with safetensors. I find current load weight implementation doubles the time cost as the tensor-model parallelism increases (refer to the belowing loading time table for LLaMA-65B).

Parallelism Degree| Original (minutes) | Safetensors (minutes)
---|---|---
1| ~5 | ~5
2| ~10 | ~5
4| ~10 | ~5

I think it is ready for review. 
Code adapted from https://github.com/huggingface/text-generation-inference/blob/v0.8.2/server/text_generation_server/models/flash_llama.py#L206

### CODE REVIEW

**Review by AlpinDale — APPROVED** [review]

**Comment by AlpinDale** [comment]
Oh sorry, didn't mean to do that.

**Comment by lucasjinreal** [comment]
@AlpinDale Can merge this? Currently model loading are extremly slow

**Comment by creatorrr** [comment]
@JF-D could you please add some comments to your changes? A tad hard to read them at the moment :grimacing:

**Comment by JF-D** [comment]
Resolve conflicts for reference.

**Review by zhuohan123 — CHANGES_REQUESTED** [review]
@JF-D Sorry for the long delay. A lot of people have actually asked for safetensor support and your PR looks great for LLaMA models! Do you think there is a possibility to extend what you did to all models by just modifying the `hf_model_weights_iterator` function?

**Comment by JF-D** [comment]
@zhuohan123 I think it's possible, and I've updated the `hf_model_weights_iterator` function. Maybe you can review it.

File: vllm/model_executor/weight_utils.py [hunk 1]
```diff
@@ -3,9 +3,11 @@
 import glob
 import json
 import os
+from collections import defaultdict
 from typing import Iterator, List, Optional, Tuple
 
 from huggingface_hub import snapshot_download
+from safetensors.torch import load_file, save_file
```
Comment (AlpinDale) [inline vllm/model_executor/weight_utils.py:10]: We also want to import `safe_open` from `safetensors.torch` here.

File: vllm/model_executor/weight_utils.py [hunk 2]
```diff
@@ -71,16 +161,45 @@ def hf_model_weights_iterator(
...
+    if not isinstance(loaded_weight, torch.Tensor):
+        loaded_weight = loaded_weight[:]
+
+    param[:loaded_weight.shape[0]].copy_(loaded_weight)
+
+
 def load_tensor_parallel_weights(
     param: torch.Tensor,
-    loaded_weight: torch.Tensor,
+    loaded_weight: torch.Tensor or object,
```
Comment (zhuohan123) [inline vllm/model_executor/weight_utils.py:202]: Style nit:
```suggestion
    loaded_weight: Any,  # `torch.Tensor` or `PySafeSlice`
```
If you can find the definition of `PySafeSlice` class, please use `PySafeSlice` here directly.

**Review by zhuohan123 — CHANGES_REQUESTED** [review]
Again, thank you for your great contribution! We tested the code last week and it works great! Left some questions and small suggestions. Once fixed, this PR should be able to be merged. BTW you can run `format.sh --all` to format your changes.

**Review by zhuohan123 — APPROVED** [review]
Thank you for your contribution again! I will merge this PR first, and then add safetensor loading for other models in another PR.

**Pull request merged**